src/software/new_geom/polynomial.h

```diff
@@ -0,0 +1,156 @@
+#pragma once
+
+#include <cmath>
+#include <vector>
+
+/**
+ * Polynomial is a representation of a polynomial
+ * that can calculate values at a given input
+ */
+class Polynomial
+{
+   public:
+    // Threshold for coefficient comparison.
+    // http://www.cplusplus.com/forum/beginner/95128/
+    static constexpr double EPSILON = 1e-15;
+
+    /**
+     * Construct a zero polynomial
+     */
+    explicit Polynomial();
+
+    /**
+     * Construct a polynomial from coefficients of the form
+     * coeffs[0] + coeffs[1]*x^(1) + ... + coeffs[n-1]*x^(n-1)
+     *
+     * @param coeffs coefficients of the polynomial
+     */
+    explicit Polynomial(const std::vector<double> &coeffs);
+
+    /**
+     * Construct a polynomial from coefficients of the form
+     * coeffs[0] + coeffs[1]*x^(1) + ... + coeffs[n-1]*x^(n-1)
+     *
+     * @param coeffs coefficients of the polynomial
+     */
+    explicit Polynomial(const std::initializer_list<double> &coeffs);
+
+    /**
+     * Construct a linear polynomial from two pairs of input/output
+     *
+     * @param constraint1 first pair of values: first is the input, second is the output
+     * @param constraint2 second pair of values: first is the input, second is the output
+     *
+     * @throws std::invalid_argument if constraint1.first == constraint2.first
+     */
+    explicit Polynomial(const std::pair<double, double> &constraint1,
+                        const std::pair<double, double> &constraint2);
```

Comment:
I commented on this over in the `Spline` PR, but I still disagree with having a function interface where it's trivial for the user to pass in entirely invalid values (ex. `{0,1}, {0,2}`). I'm also not sure that there's really a need for a special case method for 1st order polynomials. Thoughts on just having the user call the constructor with coefficients? (@jonathanlew - FYI since we were discussing this over on the `Spline` PR).

Reply:
This was made for convenience when constructing splines from points. We should have a polynomial spline constructor that interpolates points with given degree polynomials and uses right number continuity and arbitrary boundary conditions to determine coefficients. Then we can use the coefficient constructor

Reply:
If this stays in new_geom until we add the function I described, then this constructor can be removed

Reply:
Removed in af20cdc76f49a87c92abe851d79b90511c7bfb33